numpyro/infer/mixed_hmc.py

```diff
@@ -0,0 +1,220 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import namedtuple
+from functools import partial
+
+from jax import jacfwd, grad, lax, ops, random
+import jax.numpy as jnp
+
+from numpyro.infer.hmc import momentum_generator
+from numpyro.infer.hmc_gibbs import DiscreteHMCGibbs
+from numpyro.infer.hmc_util import euclidean_kinetic_energy, warmup_adapter
+from numpyro.util import cond, fori_loop, identity, ravel_pytree
+
+
+MixedHMCState = namedtuple("MixedHMCState", "z, hmc_state, rng_key, accept_prob")
+
+
+class MixedHMC(DiscreteHMCGibbs):
+    """
+    Implementation of Mixed Hamiltonian Monte Carlo (reference [1]).
+
+    .. note:: The number of discrete sites to update at each MCMC iteration
+        (`n_D` in reference [1]) is fixed at value 1.
+
+    **References**
+
+    1. *Mixed Hamiltonian Monte Carlo for Mixed Discrete and Continuous Variables*,
+       Guangyao Zhou (2020)
+    2. *Peskun's theorem and a modified discrete-state Gibbs sampler*,
+       Liu, J. S. (1996)
+
+    :param inner_kernel: A :class:`~numpyro.infer.hmc.HMC` kernel.
+    :param int num_discrete_updates: Number of times to update discrete variables.
+        Defaults to the number of discrete latent variables.
+    :param bool random_walk: If False, Gibbs sampling will be used to draw a sample from the
```

Comment:
nit: ... conditional `p(gibbs_site | remaining sites)`, where `gibbs_site` is one of the discrete sample sites in the model.